### ISSUE TITLE
The constant-memory Neural ODE adjoint needs the missing numerical-accuracy caveat

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#25]
## Finding

At commit `af3f866dea23e38c89e88f485bce99ff546d93a9`, the Neural ODE annotation says:

> “The adjoint sensitivity method for continuous-depth networks: backpropagate by solving a second ODE backwards in time, at constant memory.”

Source: [`bibliography.html` lines 1431–1434](https://github.com/microprediction/latentreasoning/blob/af3f866dea23e38c89e88f485bce99ff546d93a9/bibliography.html#L1431-L1434).

This accurately describes the original proposal’s memory claim, but presents it as an unqualified backward pass. The constant-memory implementation is not generally an accurate gradient of the numerical forward solve. The bibliography omits the primary correction literature, especially ANODE, even though that limitation is central to the section’s comparison with implicit/equilibrium gradients.

## What the original method actually needs

The continuous adjoint equation depends on the entire forward trajectory `z(t)`. To avoid storing it, the original Neural ODE implementation keeps the terminal state and reconstructs `z(t)` while integrating an augmented system backward.

Chen et al. explicitly describe this constant-memory construction in [§2](https://arxiv.org/html/1806.07366#S2). With an exact flow and exact arithmetic, the continuous sensitivity equations are valid. A numerical ODE solver, however, supplies neither condition automatically.

There are two separate implementation errors:

1. **Reverse reconstruction can be ill-conditioned or fail.** A forward-stable dissipative flow contracts perturbations; reversing it exponentially amplifies numerical error. The adjoint then evaluates Jacobians along the wrong reconstructed path.
2. **Optimize-then-discretize can disagree with discretize-then-optimize.** The continuous adjoint is not generally the exact reverse-mode derivative of the particular adaptive/discrete solver that produced the forward value.

Thus “solve a second ODE backwards” is an O(1)-memory gradient estimator, not a universal exact backpropagation recipe.

## Primary correction omitted by the site

Gholami, Keutzer & Biros, [*ANODE: Unconditionally Accurate Memory-Efficient Gradients for Neural ODEs*](https://arxiv.org/abs/1902.10298), directly analyzes the original method. It reports:

- reverse ODE solves can be numerically unstable for ReLU/non-ReLU activations and convolutional operators;
- reconstruction errors can produce order-one gradient errors;
- adaptive timestepping does not in general remove the instability;
- continuous and discrete gradients can be inconsistent enough to cause divergent training.

See [the contributions and failure analysis](https://arxiv.org/html/1902.10298#S1) and [the reverse-stability examples](https://arxiv.org/html/1902.10298#S3).

ANODE’s accurate alternative checkpoints forward states and uses discretize-then-optimize differentiation. Its basic memory is `O(L) + O(N_t)`, not constant in the number of solver steps; further checkpointing reduces memory only by spending extra recomputation.

Later methods preserve the tradeoff rather than making it disappear. For example, Zhuang et al., [*MALI: A memory efficient and reverse accurate integrator for Neural ODEs*](https://arxiv.org/abs/2102.04668), introduces a specially designed reversible integrator precisely because ordinary adjoint implementations can have inaccurate reverse trajectories. That architecture-specific remedy should not be confused with a guarantee for an arbitrary black-box ODE solve.

## Explicit stability example

For the scalar ODE

```
dz/dt = -lambda z,  lambda > 0,
```

the forward map over time `T` multiplies by `exp(-lambda T)`, so forward errors contract. Reconstructing backward multiplies perturbations by `exp(lambda T)`. A terminal rounding error `epsilon` therefore becomes `epsilon exp(lambda T)`. No finite-precision constant-memory reconstruction is uniformly stable as `lambda T` grows.

The ANODE paper gives this example quantitatively and reports that `lambda=-100` needs roughly 200,000 steps to reverse to 1% accuracy over unit time, while larger magnitudes become impossible in double precision.

## Consequence

The current annotation makes the continuous-time method sound like a clean analogue of the equilibrium backward pass. Both avoid storing an unrolled trajectory in their ideal formulas, but their numerical failure modes differ:

- a DEQ backward solve requires a well-conditioned, nonsingular equilibrium Jacobian;
- a Neural ODE adjoint additionally needs an accurate forward trajectory during the reverse solve, and the constant-memory reconstruction can be unstable or inconsistent with the discrete forward solver.

That distinction matters when importing either method into recurrent latent reasoning.

## Concrete correction

A source-faithful replacement would be:

> “Neural ODE proposed an adjoint solve with memory constant in the number of integration steps by reconstructing the forward trajectory backward in time. Later work showed that this reconstruction can be numerically unstable and that the continuous adjoint need not equal the gradient of the discrete forward solver. Accurate alternatives use checkpointing/recomputation or specially reversible integrators, trading memory, compute, or architectural constraints.”

Add ANODE, and optionally MALI or a comparable discrete-adjoint/checkpointing result, beside the original Neural ODE citation.

**Comment by microprediction** [comment]
Correct. The constant-memory version reconstructs the forward trajectory by integrating backwards, so it is not in general an exact gradient of the numerical forward solve. Added that caveat and cited ANODE (arXiv:1902.10298, verified) alongside it.

Fixed in 0c96d71.
